### ISSUE TITLE
GitHub Storage: session string exposed

### ISSUE THREAD

**Issue by Keyrxng** [ubiquity-os-marketplace/ubiquity-os-kernel-telegram#12]
Relates to #3

The config repo is now public and GitHub layer storage is supposed to push into that repo.

This plugin has one ultra sensitive env var which is the MTProtoAPI Session String.

If that is stored in a public repo as raw text the TG can be hijacked easily.

1. Add encryption/decryption
2. Find another way to store just that var
3. Adjust the storage layer solution that has been presented as-is

_original [context](https://github.com/ubiquity-os-marketplace/ubiquity-os-kernel-telegram/pull/3#issuecomment-2432743307)_

**Comment by Keyrxng** [comment]
This is an urgent task as it's a blocker now for #3 and estimate it to be about 4 hours+ of work depending on the approach to be implemented

**Comment by Keyrxng** [comment]
rfc @0x4007

**Comment by 0x4007** [comment]
We can just use our existing X25519 encryption @rndquu rfc

**Comment by Keyrxng** [comment]
You mentioned during review that we'll make Global Storage private and handle it that way but yes we could use the same encryption we use in other projects if that's the method to implement.

**Comment by rndquu** [comment]
@Keyrxng 

Trying to understand how the whole plugin works.

1. A new telegram bot (let's say `ubiquity-os-telegram-bot`) is created
2. Contributor opens `ubiquity-os-telegram-bot` and prints `/start`
3. `ubiquity-os-telegram-bot` receives a webhook from the user where we save username and user id to some storage (supabase/github json/whatever...)
4. Contributor solves a github issue, issue is closed as "completed", `ubiquity-os` github app receives a webhook
5. `ubiquity-os` github app retrieves contributor's telegram username with user id and (using `ubiquity-os-telegram-bot` credentials) sends a direct message in telegram to a user that issue has been resolved

Does it work this way?

**Comment by Keyrxng** [comment]
The session string is for the MTProto API which is responsible for the `workrooms` feature, it doesn't affect notifications as that's a Bot API feature.

1. Interface to come to allow plugins to ping this worker and send an arbitrary msg from any of our plugins. Imagine Security Monitoring implemented, we use these DMs to notify team etc.
2. Plugin devs are quota'd for rate limit; we build analytics into `plugin-template` feed it back to the kernel, pass it off to a dedicated worker which tracks usage in real-time across all incoming plugins/org/developer etc and we are able to message all parties if something is wrong or about to break.
3. Able to post `.fatal()` errors from the logger straight into a dedicated core team TG channel

There are a lot more applications for this but after #13 is fully installed which will be later today I'll make a task and aim to really improve the docs etc for this plugin as it is a little complex

**Comment by Keyrxng** [comment]
The session string solve is easy:

1. We encrypt with same approach we use elsewhere, I'll write another setup CLI for this process.
2. We use a private storage location, so a private repo. <-- this solves for more than just my use-case and I can't see the downsides of it personally.

**Comment by rndquu** [comment]
@Keyrxng 

1. `ubiquity-os-kernel-telegram` is a telegram bot, not a github app, correct?
2. `ubiquity-os-kernel-telegram` is also able to receive webhooks from https://github.com/ubiquity-os/ubiquity-os-kernel, correct?
3. Is the flow in the following picture correct?
<img width="2192" alt="flow" src="https://github.com/user-attachments/assets/8469f222-556e-425a-b737-fea233a454de">

**Comment by Keyrxng** [comment]
1. It's not a GitHub app but requires usage of one for an access token (and writes to private storage repos later)
2. It's a kernel plugin in the traditional sense yes, same as `start-stop` or any other.
3. Yeah that's it. Kernel receives and forwards as normal; all logic along that path has full kernel input. The clear difference is; this plugin receives payloads from a source other than the kernel and along that path it has zero kernel input (yml config settings, etc...), which is why we need a token and smart usage of TypeBox schema defaults.

---

Separate from that you need to also consider that the Telegram side has two "branches" we'll call them:

- MTProto API: personal account to create workrooms, it can DM without a previous connection. It runs via actions only and has nothing to do with the BotFather. It needs the session string.
- Bot API: you know this well already but it runs via the worker; could likely use it within actions but not the other way around

**Comment by rndquu** [comment]
> MTProto API: personal account to create workrooms, it can DM without a previous connection. It runs via actions only and has nothing to do with the BotFather. It needs the session string.

Why can't it be run in a cloudflare worker?

**Comment by Keyrxng** [comment]
Basically yes.

1. The worker receives payloads from both the kernel so it can parse comments, see who got disqualified etc and telegram so it can handle slash commands via telegram. We can send messages, remove from chat, handle forums all of that.
2. As an action solely for the workroom feature (and any future MTProto requiring features)

**Comment by rndquu** [comment]
> So I think we should go back to https://github.com/ubiquity-os-marketplace/ubiquity-os-kernel-telegram/issues/20

@Keyrxng So can we close current issue as "not planned" in favor of https://github.com/ubiquity-os-marketplace/ubiquity-os-kernel-telegram/issues/20?

**Comment by Keyrxng** [comment]
> > So I think we should go back to #20
> 
> @Keyrxng So can we close current issue as "not planned" in favor of #20?

No, #20 fixes the dual config entry and allows the worker to dispatch workflows to run the `workrooms` feature, so that it's self-contained and only has one config entry in `ubiquityos-config.yaml`

@0x4007 wants DB deps removed and GH storage used instead. I need to store a session string in (currently) a public repo. So either:

1. We have a private storage repo
2. We encrypt/decrypt values from the SDK (means GH storage needs built from the SDK rather than my single minded approach for this tg plugin
3. Something else

**Comment by 0x4007** [comment]
2. Using our X25519 key. I think @eresdev is also doing something similar with personal access token encryption for the "personal agents" feature.

**Comment by 0x4007** [comment]
I don't think anybody is working on it 

The final deliverable should be in the sdk

**Comment by Keyrxng** [comment]
> I don't think anybody is working on it

I had made it a priority because you have been talking about it for literally months without any progress on it which is why I threw an implementation together within my plugin just before I was removed from the team.

I expect I'll have standard contributor task limits these days and can only work on `Priority 3` priced tasks also, so I'll just have to chop away at what I'm able to work on in free time I guess with the new shift
